File: personalizatio-sdk/src/main/java/com/personalizatio/stories/views/StoriesView.java
package com.personalizatio.stories.views;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.database.ContentObserver;
import android.media.AudioManager;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.personalizatio.Api;
import com.personalizatio.OnLinkClickListener;
import com.personalizatio.R;
import com.personalizatio.SDK;
import com.personalizatio.stories.Player;
import com.personalizatio.stories.Settings;
import com.personalizatio.stories.StoriesUtils;
import com.personalizatio.stories.viewAdapters.StoriesAdapter;
import com.personalizatio.stories.models.Story;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

final public class StoriesView extends ConstraintLayout implements StoriesAdapter.ClickListener, SDK.ShowStoriesRequestListener {

	private StoriesAdapter adapter;
	private final List<Story> list = new ArrayList<>();
	private ContentObserver observer;
	private final Settings settings = new Settings();
	private String code;
	private Player player;
	@Nullable private OnLinkClickListener clickListener;
	private boolean mute = true;
	private Runnable muteListener;

	public StoriesView(Context context, String code) {
		super(context);
		this.setCode(code);
		initialize();
	}

	public StoriesView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		parseAttrs(attrs);
	}

	public StoriesView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		parseAttrs(attrs);
	}

	@TargetApi(Build.VERSION_CODES.LOLLIPOP)
	public StoriesView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);
		parseAttrs(attrs);
	}

	public Settings getSettings() {
		return settings;
	}

	public Player getPlayer() {
		return player;
	}

	public void setPlayer(Player player) {
		this.player = player;
	}

	/**
	 * Вызывать, когда объект сторисов удален с экрана и больше не нужен
	 */
	public void release() {
		getPlayer().release();
	}

	@Override
	protected void onDetachedFromWindow() {
		super.onDetachedFromWindow();
		unregisterObserver();
	}

	@Override
	protected void onFinishInflate() {
		super.onFinishInflate();
		initialize();
	}

	private void parseAttrs(AttributeSet attrs) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.StoriesView);
		setCode(typedArray.getString(R.styleable.StoriesView_code));
	}

	//Инициализация
	private void initialize() {
		View view = inflate(getContext(), R.layout.stories, this);
		RecyclerView stories = view.findViewById(R.id.stories);

		adapter = new StoriesAdapter(this, list, this);
		stories.setAdapter(adapter);

		Handler handler = new Handler(Looper.getMainLooper()) {
			@Override
			public void handleMessage(Message msg) {
				registerObserver();
				adapter.notifyDataSetChanged();
			}
		};

		//Плеер для просмотра видео
		setPlayer(new Player(getContext()));

		getSettings().failed_load_text = getResources().getString(R.string.failed_load_text);

		//Запрашиваем сторисы
		SDK.stories(getCode(), new Api.OnApiCallbackListener() {
			@Override
			public void onSuccess(JSONObject response) {
				Log.d("stories", response.toString());
				var stories = StoriesUtils.GetStories(response);

				if(!stories.isEmpty()) {
					handler.sendEmptyMessage(1);

					list.addAll(stories);
				}
			}
		});

		SDK.setShowStoriesRequestListener(this);
	}

    @Override
	public void onStoryClick(int index) {
		Story story = list.get(index);
		story.resetStartPosition();

		showStories(list, index, () -> adapter.notifyDataSetChanged());
	}

	/**
	 * Устанавливает слушатель клика по элементам
	 * @param listener OnLinkClickListener
	 */
	public void setOnLinkClickListener(@Nullable OnLinkClickListener listener) {
		this.setClickListener(listener);
	}

	public void muteVideo(boolean mute) {
		this.mute = mute;
	}

	public boolean isMute() {
		return mute;
	}

	public void registerObserver() {
		observer = new ContentObserver(new Handler()) {
			@Override
			public void onChange(boolean selfChange) {
				super.onChange(selfChange);
				AudioManager manager = (AudioManager) getContext().getSystemService(Context.AUDIO_SERVICE);
				if( Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ) {
					mute = manager.isStreamMute(AudioManager.STREAM_MUSIC);
					if (muteListener != null) {
						muteListener.run();
					}
				}
			}
		};
		getContext().getContentResolver().registerContentObserver(android.provider.Settings.System.CONTENT_URI, true, observer);
	}

	public void unregisterObserver() {
		if( observer != null ) {
			getContext().getContentResolver().unregisterContentObserver(observer);
		}
	}

	public String getCode() {
		return code;
	}

	public void setCode(String code) {
		this.code = code;
	}

	@Nullable
	public OnLinkClickListener getClickListener() {
		return clickListener;
	}

	public void setClickListener(@Nullable OnLinkClickListener clickListener) {
		this.clickListener = clickListener;
	}

	public void setMuteListener(Runnable muteListener) {
		this.muteListener = muteListener;
	}

	@Override
	public void onShowStoriesRequest(List<Story> stories) {
		Handler handler = new Handler(getContext().getMainLooper());
		handler.post(() -> showStories(stories, 0, () -> {}));
	}

	private void showStories(List<Story> stories, int startPosition, Runnable completeListener) {
		StoryDialog dialog = new StoryDialog(this, stories, startPosition, completeListener);
		dialog.show();
	}
}

File: personalizatio-sdk/src/main/kotlin/com/personalizatio/stories/views/StoriesView.kt
package com.personalizatio.stories.views

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.database.ContentObserver
import android.media.AudioManager
import android.os.Build
import android.os.Handler
import android.util.AttributeSet
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.personalizatio.OnLinkClickListener
import com.personalizatio.R
import com.personalizatio.SDK
import com.personalizatio.stories.Player
import com.personalizatio.stories.Settings
import com.personalizatio.stories.models.Story
import com.personalizatio.stories.viewAdapters.StoriesAdapter
import com.personalizatio.stories.viewAdapters.StoriesAdapter.ClickListener

class StoriesView : ConstraintLayout, ClickListener {
    private var adapter: StoriesAdapter? = null
    private val stories: MutableList<Story> = ArrayList()
    private var observer: ContentObserver? = null
	val settings: Settings = Settings()
	var code: String? = null
	var player: Player? = null
    var clickListener: OnLinkClickListener? = null
    var isMute: Boolean = true
        private set
    var muteListener: Runnable? = null

    constructor(context: Context?, code: String?) : super(context!!) {
        this.code = code
        initialize()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context!!, attrs) {
        parseAttrs(attrs)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context!!, attrs, defStyleAttr) {
        parseAttrs(attrs)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int)
            : super(context!!, attrs, defStyleAttr, defStyleRes) {
        parseAttrs(attrs)
    }

    @SuppressLint("NotifyDataSetChanged")
    internal fun updateStories(stories: List<Story>) {
        this.stories.clear()
        this.stories.addAll(stories)

        Handler(context.mainLooper).post {
            registerObserver()
            adapter?.notifyDataSetChanged()
        }
    }

    /**
     * Вызывать, когда объект сторисов удален с экрана и больше не нужен
     */
    fun release() {
        player?.release()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        unregisterObserver()
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        initialize()
    }

    private fun parseAttrs(attrs: AttributeSet?) {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.StoriesView)
        this.code = typedArray.getString(R.styleable.StoriesView_code)
    }

    //Инициализация
    private fun initialize() {
        val view = inflate(context, R.layout.stories, this)
        val stories = view.findViewById<RecyclerView>(R.id.stories)

        adapter = StoriesAdapter(this, this.stories, this)
        stories.adapter = adapter

        //Плеер для просмотра видео
        player = Player(context)

        settings.failed_load_text = resources.getString(R.string.failed_load_text)
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onStoryClick(index: Int) {
        val story = stories[index]

        story.resetStartPosition()

        showStories(stories, index) { adapter?.notifyDataSetChanged() }
    }

    fun muteVideo(mute: Boolean) {
        this.isMute = mute
    }

    fun registerObserver() {
        observer = object : ContentObserver(Handler()) {
            override fun onChange(selfChange: Boolean) {
                super.onChange(selfChange)
                val manager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    isMute = manager.isStreamMute(AudioManager.STREAM_MUSIC)
                    if (muteListener != null) {
                        muteListener!!.run()
                    }
                }
            }
        }
        context.contentResolver.registerContentObserver(
            android.provider.Settings.System.CONTENT_URI,
            true,
            observer as ContentObserver
        )
    }

    fun unregisterObserver() {
        if (observer != null) {
            context.contentResolver.unregisterContentObserver(observer!!)
        }
    }

    internal fun showStory(storyId: Int): Boolean {
        for (story in stories) {
            if (storyId == story.id) {
                showStory(story)
                return true
            }
        }

        return false
    }

    internal fun showStory(story: Story) {
        story.startPosition = 0

        val stories = ArrayList<Story>(1)
        stories.add(story)

        val handler = Handler(context.mainLooper)
        handler.post {
            showStories(stories, 0) {
                story.startPosition = 0
            }
        }
    }

    private fun showStories(stories: List<Story>, startPosition: Int, completeListener: Runnable) {
        val dialog = StoryDialog(this, stories, startPosition, completeListener)
        dialog.show()
    }
}
